packages/form-core/src/utils.ts

```diff
@@ -376,3 +380,30 @@ export function shallow<T>(objA: T, objB: T) {
   }
   return true
 }
+
+/**
+ * Determines the logic for determining the error value to set on the field meta within the form sync validation.
+ * @private
+ */
+export const determineErrorValue = ({
+  newFormValidatorError,
+  isPreviousErrorFromFormValidator,
+  previousErrorValue,
+}: {
+  newFormValidatorError: ValidationError
+  isPreviousErrorFromFormValidator: boolean
+  previousErrorValue: ValidationError
+}) => {
+  // All falsy values are not considered errors
+  if (newFormValidatorError) {
+    return { newErrorValue: newFormValidatorError, newSource: 'form' }
+  }
+
+  // Clears form level error since it's now stale
+  if (isPreviousErrorFromFormValidator) {
+    return { newErrorValue: undefined, newSource: undefined }
+  }
+
+  // Do not clear field level validators since they're handled by the field validator
+  return { newErrorValue: previousErrorValue, newSource: 'field' }
```

Comment:
I think there's an extra case here, if `previousErrorValue` is `undefined` we should probably also return `newSource: undefined`